### ISSUE TITLE
Tools to create Tables

### ISSUE THREAD

**Issue by Vijay-Nirmal** [neilsustc/vscode-markdown#23]
- Code snippet to create basic table layout.
    ```
    |     |     |
    | --- | --- |
    |     |     |
    ```
   Pressing Tab should loop the cursor in different cells
- Right clicking inside the table should have options to Add New Row At End, Add New Column At End, Add New Row(Creates new Row next to the current row), Add New Column (Creates new Column next to the current Column), Delete Row(Deletes current Row), Delete Column(Deletes current Column).

**Comment by neilsustc** [comment]
Thanks for your suggestion.
It is a good feature. But I would like to think more about the UX. Right clicking to add row/column is kind of... "strange"

**Comment by Vijay-Nirmal** [comment]
What about Shortcut keys? But I would prefer Right Click because it is more visible to users.
Will VS code API allows you to make a ToolBar?

**Comment by neilsustc** [comment]
vscode API provides very limited support for these customizations. That is why I am hesitant.

**Comment by Vijay-Nirmal** [comment]
New Method. When the user places the cursor in a specific cell
- Pressing Double Enter creates new Row
- Pressing Double Tab creates new Column
- Pressing Ctrl + Double Enter Deletes current Row
- Pressing Ctrl + Double Tab Deletes current Column

**Comment by neilsustc** [comment]
I'm afraid that we cannot easily detect double enter/tab event… (how do we do when we find the enter key is pressed once? Breaking line immediately? Or waiting for another pressing?)

Thanks again for your suggestion.

And before I find a good way, you can add some simple shortcuts for workaround.

**Comment by Vijay-Nirmal** [comment]
Then use
- Ctrl + Enter(Tab) for Create
- Shift + Enter(Tab) for Delete

**Comment by neilsustc** [comment]
Let me do some investigations on how other Markdown editors do this

**Comment by neilsustc** [comment]
[Caret](https://caret.io/) is a good Markdown editor. It

- creates tables on <kbd>Enter</kbd> based on header `| abc | abc |`
- creates table rows on <kbd>Enter</kbd>
- resizes table columns on typing

![table-iloveimg-compressed](https://user-images.githubusercontent.com/7588612/28926824-865f771e-789b-11e7-9d69-f9802e864da5.gif)

**Comment by Vijay-Nirmal** [comment]
Creating columns?

**Comment by neilsustc** [comment]
Caret cannot do that.

[Typora](https://typora.io/) has a custom control to change table size and alignment. But it is hard to bring that to vscode.

![table-iloveimg-compressed 1](https://user-images.githubusercontent.com/7588612/28947210-4c5d5ec8-78e1-11e7-850c-ddd73054ad34.gif)

**Comment by Vijay-Nirmal** [comment]
What about adding additional column in header automatically creates column in each row?

**Comment by neilsustc** [comment]
I agree that adding table body as typing table header is a natural way.
But some tricks are needed to implement this (because of vscode API). So at this time I'm not motivated to do it 😓.

As for the alignment, choose `:---`, `---:` and `:---:` in the second line.

**Comment by yume-chan** [comment]
I think you can use CodeLens and CodeAction to provide these actions.

Put the cursor in a cell, then the CodeLens provides `align left`, `align center`, `align right`, `insert column left`, `insert column right` commands, while the CodeAction provides `insert row above` and `insert row below` commands.

**Comment by neilsustc** [comment]
Of course you can do that. But the problem is it will add too many "buttons" in the UI making it looks bad.

**Comment by doggy8088** [comment]
@neilsustc Is it possible to add **Convert tabular data to tables** feature?  So that I can easily copy Excel data into a Markdown table.

**Comment by neilsustc** [comment]
Accessing clipboard is a big problem. More info [here](https://github.com/neilsustc/vscode-markdown/issues/129#issuecomment-387208188)

> The VSCode API doesn't provide methods to access the clipboard. There are several open issues regarding this issue (217, 4972, lots of duplicates), but it doesn't look like Microsoft has plans to implement an API for the clipboard. This could be an issue as there are several issues noted by users with the workaround libraries recommended by the VSCode team.

**Comment by doggy8088** [comment]
@neilsustc Maybe you can check this out: [Markdown Shortcuts](https://marketplace.visualstudio.com/items?itemName=mdickin.markdown-shortcuts).  It do have a **Convert tabular data to tables** feature in it.

**Comment by neilsustc** [comment]
OK, I understand now. (I thought you meaned to paste Markdown table directly from Excel data)

I opened another issue for it. You can use Markdown Shortcuts as a complement before that feature is added.

**Comment by pellejacobs** [comment]
In case this feature is still being considered, Atom has an amazing package that is very intuitive to use: https://atom.io/packages/markdown-table-editor. 

Shotcuts like `tab` to jump to a next cell are really useful.

**Comment by yzhang-gh** [comment]
Thanks for the information. This is similar to Caret I mentioned before

> [Caret](https://caret.io/) is a good Markdown editor. It
>
> - creates tables on <kbd>Enter</kbd> based on header `| abc | abc |`
> - creates table rows on <kbd>Enter</kbd>
> - [added] goes to next cell on <kbd>Tab</kbd>
> - resizes table columns on typing/<kbd>Tab</kbd>
>
> ![table-iloveimg-compressed](https://user-images.githubusercontent.com/7588612/28926824-865f771e-789b-11e7-9d69-f9802e864da5.gif)

These are completely doable although also require much effort.

**Comment by pomdtr** [comment]
This extension kind of does the job : https://marketplace.visualstudio.com/items?itemName=JayFiDev.tablegenerator&ssr=false#review-details
